crates/blockchain/payload.rs

```diff
@@ -214,21 +237,57 @@ pub fn build_payload(
     debug!("Building payload");
     let mut evm_state = evm_state(store.clone(), payload.header.parent_hash);
     let mut context = PayloadBuildContext::new(payload, &mut evm_state);
-    apply_withdrawals(&mut context)?;
-    fill_transactions(&mut context)?;
-    finalize_payload(&mut context)?;
+    let mut block_cache = apply_withdrawals(&mut context)?;
+    fill_transactions(&mut context, &mut block_cache)?;
+    finalize_payload(&mut context, &mut block_cache)?;
+
     Ok((context.blobs_bundle, context.block_value))
 }
 
-pub fn apply_withdrawals(context: &mut PayloadBuildContext) -> Result<(), EvmError> {
-    // Apply withdrawals & call beacon root contract, and obtain the new state root
-    let spec_id = spec_id(&context.chain_config()?, context.payload.header.timestamp);
-    if context.payload.header.parent_beacon_block_root.is_some() && spec_id == SpecId::CANCUN {
-        beacon_root_contract_call(context.evm_state, &context.payload.header, spec_id)?;
+pub fn apply_withdrawals(
+    context: &mut PayloadBuildContext,
+) -> Result<HashMap<Address, Account>, EvmError> {
+    #[cfg(feature = "levm")]
+    {
+        let mut block_cache = HashMap::new();
+        // Apply withdrawals & call beacon root contract, and obtain the new state root
+        let spec_id = spec_id(&context.chain_config()?, context.payload.header.timestamp);
+        if context.payload.header.parent_beacon_block_root.is_some() && spec_id == SpecId::CANCUN {
+            let store_wrapper = Arc::new(StoreWrapper {
+                store: context.evm_state.database().unwrap().clone(),
+                block_hash: context.payload.header.parent_hash,
+            });
+            let report = beacon_root_contract_call_levm(
+                store_wrapper.clone(),
+                &context.payload.header,
+                spec_id,
+            )?;
+
+            let mut new_state = report.new_state.clone();
+
+            // Now original_value is going to be the same as the current_value, for the next transaction.
+            // It should have only one value but it is convenient to keep on using our CacheDB structure
+            for account in new_state.values_mut() {
+                for storage_slot in account.storage.values_mut() {
+                    storage_slot.original_value = storage_slot.current_value;
+                }
+            }
+
+            block_cache.extend(new_state);
+        }
+        Ok(block_cache)
+    }
+    #[cfg(not(feature = "levm"))]
+    {
+        // Apply withdrawals & call beacon root contract, and obtain the new state root
+        let spec_id = spec_id(&context.chain_config()?, context.payload.header.timestamp);
+        if context.payload.header.parent_beacon_block_root.is_some() && spec_id == SpecId::CANCUN {
```

Comment:
Here we have to take into account future forks:
```suggestion
        if context.payload.header.parent_beacon_block_root.is_some() && spec_id >= SpecId::CANCUN {
```

Reply:
well, you just updated it to `Fork`, but the idea is the same.

Reply:
Nice!, done here [b3da929](https://github.com/lambdaclass/ethrex/pull/1817/commits/b3da929c4b605e33eb0989dd8ea7d6157bc9625f)